File: phantom-mac-app/Sources/Phantom/Services/ProviderResiliencePolicy.swift
import Foundation

enum ProviderFailureKind: String {
    case rateLimited = "rate_limited"
    case authentication = "authentication_failed"
    case transient = "provider_transient"
    case terminal = "provider_error"
    case cancelled
}

struct ProviderFailureDecision {
    let kind: ProviderFailureKind
    let canRotateCredential: Bool
    let cooldown: TimeInterval
}

enum ProviderResiliencePolicy {
    static let managedDesktopMaxAttempts = 1
    static let byoDesktopMaxAttempts = 2

    static func classify(statusCode: Int? = nil, message: String = "") -> ProviderFailureDecision {
        let value = message.lowercased()
        if statusCode == 429 || contains(value, ["rate_limit", "ratelimit", "too many requests", "quota", "resource_exhausted"]) {
            return ProviderFailureDecision(kind: .rateLimited, canRotateCredential: true, cooldown: 5 * 60)
        }
        if statusCode == 401 || statusCode == 403 || contains(value, ["invalid api key", "invalid key", "unauthorized", "authentication"]) {
            return ProviderFailureDecision(kind: .authentication, canRotateCredential: true, cooldown: 24 * 60 * 60)
        }
        if contains(value, ["cancelled", "canceled"]) {
            return ProviderFailureDecision(kind: .cancelled, canRotateCredential: false, cooldown: 0)
        }
        if [408, 500, 502, 503, 504].contains(statusCode ?? 0)
            || contains(value, ["timeout", "timed out", "overloaded", "capacity", "network", "transport", "connection", "stream ended unexpectedly", "empty response"]) {
            return ProviderFailureDecision(kind: .transient, canRotateCredential: true, cooldown: 30)
        }
        return ProviderFailureDecision(kind: .terminal, canRotateCredential: false, cooldown: 0)
    }

    static func canRetry(_ failure: ProviderFailureDecision, attempt: Int, maxAttempts: Int, hasOutput: Bool) -> Bool {
        !hasOutput && attempt < maxAttempts && failure.canRotateCredential
    }

    static func canCrossLane(from: String, to: String, explicitlyOptedIn: Bool, hasOutput: Bool) -> Bool {
        from.caseInsensitiveCompare("byo") == .orderedSame
            && to.caseInsensitiveCompare("managed_extension") == .orderedSame
            && explicitlyOptedIn
            && !hasOutput
    }

    private static func contains(_ value: String, _ terms: [String]) -> Bool {
        terms.contains(where: value.contains)
    }
}

File: phantom-windows-app/Services/ProviderResiliencePolicy.cs
using System;
using System.Text.RegularExpressions;

namespace SecureOverlay.Services
{
    public enum ProviderFailureKind
    {
        RateLimited,
        Authentication,
        Transient,
        Terminal,
        Cancelled
    }

    public readonly record struct ProviderFailureDecision(
        ProviderFailureKind Kind,
        string ErrorCode,
        bool CanRotateCredential,
        TimeSpan Cooldown);

    public static class ProviderResiliencePolicy
    {
        public const int ManagedDesktopMaxAttempts = 1;
        public const int ByoDesktopMaxAttempts = 2;

        public static ProviderFailureDecision Classify(string? error)
        {
            var value = error?.Trim().ToLowerInvariant() ?? string.Empty;
            var status = Regex.Match(value, @"(?<!\d)(401|403|408|429|500|502|503|504)(?!\d)").Value;
            if (status == "429" || ContainsAny(value, "rate_limit", "ratelimit", "too many requests", "quota", "resource_exhausted"))
                return new(ProviderFailureKind.RateLimited, "rate_limited", true, TimeSpan.FromMinutes(5));
            if (status is "401" or "403" || ContainsAny(value, "invalid api key", "invalid key", "unauthorized", "authentication"))
                return new(ProviderFailureKind.Authentication, "authentication_failed", true, TimeSpan.FromHours(24));
            if (ContainsAny(value, "cancelled", "canceled"))
                return new(ProviderFailureKind.Cancelled, "cancelled", false, TimeSpan.Zero);
            if (status is "408" or "500" or "502" or "503" or "504"
                || ContainsAny(value, "timeout", "timed out", "overloaded", "capacity", "network", "transport", "connection", "stream_incomplete", "empty response"))
                return new(ProviderFailureKind.Transient, status.Length == 3 ? $"provider_{status[0]}xx" : "provider_transport_error", true, TimeSpan.FromSeconds(30));
            return new(ProviderFailureKind.Terminal, "provider_error", false, TimeSpan.Zero);
        }

        public static bool CanRetry(ProviderFailureDecision failure, int attempt, int maxAttempts, bool hasOutput)
            => !hasOutput && attempt < maxAttempts && failure.CanRotateCredential;

        public static bool CanCrossLane(string fromLane, string toLane, bool explicitlyOptedIn, bool hasOutput)
            => string.Equals(fromLane, "byo", StringComparison.OrdinalIgnoreCase)
                && string.Equals(toLane, "managed_extension", StringComparison.OrdinalIgnoreCase)
                && explicitlyOptedIn
                && !hasOutput;

        private static bool ContainsAny(string value, params string[] terms)
            => Array.Exists(terms, value.Contains);
    }
}
